### ISSUE TITLE
Failure to destroy AWS security groups with interdependent rules

### ISSUE THREAD

**Issue by vbichov** [hashicorp/terraform#5050]
This is a very simple issue. while many other issues mention security groups, I haven't found this one, so I'll explain it here.

When you define two (or more) security groups that contain security rules that each relate to the other group, Terrafrom fails to destroy these groups because each "depends" on the other.

Terraform can't even create such SGs if they are defined with rules inline.
Once you separate the creation of the groups and the rules the creation works, but Terraform doesn't know to destroy the rules before the SGs, and so fails on destroy.

**Comment by stack72** [comment]
Hi @vbichov,

Please can you supply a copy of the terraform config (minus keys) that you used, and also the resulting apply result and we can have a look at it

Thanks

Paul

**Comment by vbichov** [comment]
Hi Paul, thank you for your response.

In order to try to work around the problem I've tried using `terraform destroy -target=...` on specific rules.
to my horror, I've discovered that they have become "invisible" - terraform does not recognize that they exist, yet they still do, and can be viewed via the AWS console.

The general destroy error is:
Error applying plan:

4 error(s) occurred:

* aws_security_group.consul: DependencyViolation: resource sg-5a0c1223 has a dependent object
        status code: 400, request id:
* aws_security_group.distribution: DependencyViolation: resource sg-570c122e has a dependent object
        status code: 400, request id:
* aws_security_group.general_access: DependencyViolation: resource sg-4f0c1236 has a dependent object
        status code: 400, request id:
* aws_security_group.ansible: DependencyViolation: resource sg-540c122d has a dependent object
        status code: 400, request id:

terraform version is: 0.6.8
newer versions do not show the correct error, they just fail on timeout.

please see the relevant rules below, it is a bit much, I hope it helps:

resource "aws_security_group" "general_access" {
  name = "General_Access"
  description = "Basic access needed by most instances: SSH, Distribution and Consul - Managed by Terraform"
  vpc_id = "${aws_vpc.docker.id}"
  tags {
    Name = "General_Access"
    Farm = "${var.farm_name}"
  }
}

resource "aws_security_group_rule" "general_ansible" {
  type = "ingress"
  from_port = 22
  to_port = 22
  protocol = "6"
  source_security_group_id = "${aws_security_group.ansible.id}"

  security_group_id = "${aws_security_group.general_access.id}"
}


resource "aws_security_group_rule" "consul_agent_to_server_tcp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "egress"
  from_port = 8300
  to_port = 8300
  protocol = "tcp"
  source_security_group_id = "${aws_security_group.consul.id}"
}

resource "aws_security_group_rule" "general_consul_agent_from_agent_tcp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "ingress"
  from_port = 8301
  to_port = 8301
  protocol = "tcp"
  self = true
}

resource "aws_security_group_rule" "general_consul_agent_from_agent_udp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "ingress"
  from_port = 8301
  to_port = 8301
  protocol = "udp"
  self = true
}

resource "aws_security_group_rule" "general_consul_agent_to_agent_tcp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "egress"
  from_port = 8301
  to_port = 8301
  protocol = "tcp"
  self = true
}

resource "aws_security_group_rule" "general_consul_agent_to_agent_udp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "egress"
  from_port = 8301
  to_port = 8301
  protocol = "udp"
  self = true
}


resource "aws_security_group_rule" "general_consul_agent_from_server_tcp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "ingress"
  from_port = 8301
  to_port = 8301
  protocol = "tcp"
  source_security_group_id = "${aws_security_group.consul.id}"
}

resource "aws_security_group_rule" "general_consul_agent_from_server_udp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "ingress"
  from_port = 8301
  to_port = 8301
  protocol = "udp"
  source_security_group_id = "${aws_security_group.consul.id}"
}


resource "aws_security_group_rule" "general_consul_agent_to_server_tcp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "egress"
  from_port = 8301
  to_port = 8301
  protocol = "tcp"
  source_security_group_id = "${aws_security_group.consul.id}"
}

resource "aws_security_group_rule" "general_consul_agent_to_server_udp" {
  security_group_id = "${aws_security_group.general_access.id}"
  type = "egress"
  from_port = 8301
  to_port = 8301
  protocol = "udp"
  source_security_group_id = "${aws_security_group.consul.id}"
}

resource "aws_security_group_rule" "general_dns_tcp" {
  type = "egress"
  from_port = 53
  to_port = 53
  protocol = "6"
  source_security_group_id = "${aws_security_group.consul.id}"

  security_group_id = "${aws_security_group.general_access.id}"
}

resource "aws_security_group_rule" "general_dns_udp" {
  type = "egress"
  from_port = 53
  to_port = 53
  protocol = "17"
  source_security_group_id = "${aws_security_group.consul.id}"

  security_group_id = "${aws_security_group.general_access.id}"
}

resource "aws_security_group_rule" "general_distribution" {
  type = "egress"
  from_port = 5000
  to_port = 5000
  protocol = "6"
  source_security_group_id= "${aws_security_group.distribution.id}"

  security_group_id = "${aws_security_group.general_access.id}"
}

#----------------------------------------------------------------------------------------

resource "aws_security_group" "ansible" {
  name = "Ansible"
  description = "Warning: Can access all of the general_access security group - Managed by Terraform"
  vpc_id = "${aws_vpc.docker.id}"
  tags {
    Name = "Ansible"
    Farm = "${var.farm_name}"
  }
}
# The Ansible and Access SGs can't be created concurrently since they create a circle.
resource "aws_security_group_rule" "ansible_internal" {
  type = "egress"
  from_port = 22
  to_port = 22
  protocol = "6"
  source_security_group_id = "${aws_security_group.general_access.id}"

  security_group_id = "${aws_security_group.ansible.id}"
}

resource "aws_security_group_rule" "ansible_external" {
  type = "ingress"
  from_port = 22
  to_port = 22
  protocol = "6"
  cidr_blocks = ["0.0.0.0/0"]

  security_group_id = "${aws_security_group.ansible.id}"
}

#----------------------------------------------------------------------------------------

resource "aws_security_group" "consul" {
  name = "Consul"
  vpc_id = "${aws_vpc.docker.id}"
  tags {
    Name = "Consul"
    Farm = "${var.farm_name}"
  }
}

resource "aws_security_group_rule" "consul_serv_from_serv_tcp" {
  security_group_id = "${aws_security_group.consul.id}"
  type = "ingress"
  from_port = 8300
  to_port = 8300
  protocol = "tcp"
  self = true
}

resource "aws_security_group_rule" "consul_serv_to_serv_tcp" {
  security_group_id = "${aws_security_group.consul.id}"
  type = "egress"
  from_port = 8300
  to_port = 8300
  protocol = "tcp"
  self = true
}

resource "aws_security_group_rule" "consul_serv_from_client_tcp" {
  security_group_id = "${aws_security_group.consul.id}"
  type = "ingress"
  from_port = 8300
  to_port = 8301
  protocol = "tcp"
  source_security_group_id = "${aws_security_group.general_access.id}"
}

resource "aws_security_group_rule" "consul_serv_from_client_udp" {
  security_group_id = "${aws_security_group.consul.id}"
  type = "ingress"
  from_port = 8301
  to_port = 8301
  protocol = "udp"
  source_security_group_id = "${aws_security_group.general_access.id}"
}

resource "aws_security_group_rule" "consul_serv_to_client_tcp" {
  security_group_id = "${aws_security_group.consul.id}"
  type = "egress"
  from_port = 8301
  to_port = 8301
  protocol = "tcp"
  source_security_group_id = "${aws_security_group.general_access.id}"
}

resource "aws_security_group_rule" "consul_serv_to_client_udp" {
  security_group_id = "${aws_security_group.consul.id}"
  type = "egress"
  from_port = 8301
  to_port = 8301
  protocol = "udp"
  source_security_group_id = "${aws_security_group.general_access.id}"
}

resource "aws_security_group_rule" "consul_dns_tcp" {
  type = "ingress"
  from_port = 53
  to_port = 53
  protocol = "6"
  cidr_blocks = ["${aws_vpc.docker.cidr_block}"]

  security_group_id = "${aws_security_group.consul.id}"
}

resource "aws_security_group_rule" "consul_dns_udp" {
  type = "ingress"
  from_port = 53
  to_port = 53
  protocol = "17"
  cidr_blocks = ["${aws_vpc.docker.cidr_block}"]

  security_group_id = "${aws_security_group.consul.id}"
}

#----------------------------------------------------------------------------------------

resource "aws_security_group" "haproxy" {
  name = "HAProxy"
  vpc_id = "${aws_vpc.docker.id}"
  tags {
    Name = "HAProxy"
    Farm = "${var.farm_name}"
  }

  ingress {
    # All
    from_port = 0
    to_port = 0
    protocol = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}

#--------------------------------------------------------------------------------------

**Comment by stack72** [comment]
Hi @vbichov,

Thanks for the code here, I have been able to reproduce this and get the following output:

```
terraform destroy
Do you really want to destroy?
  Terraform will delete all your managed infrastructure.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

aws_vpc.docker: Refreshing state... (ID: vpc-1e3a217b)
aws_security_group.consul: Refreshing state... (ID: sg-42528125)
aws_security_group.general_access: Refreshing state... (ID: sg-46528121)
aws_security_group.distribution: Refreshing state... (ID: sg-43528124)
aws_security_group.ansible: Refreshing state... (ID: sg-45528122)
aws_security_group_rule.consul_serv_from_serv_tcp: Refreshing state... (ID: sgrule-4106712785)
aws_security_group_rule.consul_serv_to_serv_tcp: Refreshing state... (ID: sgrule-593258934)
aws_security_group_rule.consul_serv_to_client_tcp: Refreshing state... (ID: sgrule-2420366880)
aws_security_group_rule.consul_serv_from_client_tcp: Refreshing state... (ID: sgrule-3232666688)
aws_security_group_rule.general_consul_agent_from_agent_udp: Refreshing state... (ID: sgrule-210419073)
aws_security_group_rule.consul_serv_from_client_udp: Refreshing state... (ID: sgrule-2783393427)
aws_security_group_rule.consul_agent_to_server_tcp: Refreshing state... (ID: sgrule-4068026458)
aws_security_group_rule.general_consul_agent_from_server_udp: Refreshing state... (ID: sgrule-1450132561)
aws_security_group_rule.consul_serv_to_client_udp: Refreshing state... (ID: sgrule-1742676905)
aws_security_group_rule.general_consul_agent_to_server_udp: Refreshing state... (ID: sgrule-3961494421)
aws_security_group_rule.general_consul_agent_to_agent_tcp: Refreshing state... (ID: sgrule-1097244620)
aws_security_group_rule.general_consul_agent_from_server_tcp: Refreshing state... (ID: sgrule-3271353128)
aws_security_group_rule.general_consul_agent_to_agent_udp: Refreshing state... (ID: sgrule-3069862469)
aws_security_group_rule.general_consul_agent_to_server_tcp: Refreshing state... (ID: sgrule-461596188)
aws_security_group_rule.general_consul_agent_from_agent_tcp: Refreshing state... (ID: sgrule-2551782136)
aws_security_group_rule.general_consul_agent_from_agent_tcp: Destroying...
aws_security_group_rule.general_consul_agent_to_server_udp: Destroying...
aws_security_group.ansible: Destroying...
aws_security_group_rule.consul_agent_to_server_tcp: Destroying...
aws_security_group_rule.general_consul_agent_to_agent_udp: Destroying...
aws_security_group_rule.general_consul_agent_to_agent_tcp: Destroying...
aws_security_group_rule.general_consul_agent_from_agent_udp: Destroying...
aws_security_group_rule.general_consul_agent_to_server_tcp: Destroying...
aws_security_group_rule.consul_serv_from_client_udp: Destroying...
aws_security_group_rule.consul_serv_to_client_tcp: Destroying...
aws_security_group_rule.consul_serv_from_client_udp: Destruction complete
aws_security_group_rule.consul_serv_from_serv_tcp: Destroying...
aws_security_group_rule.general_consul_agent_from_agent_tcp: Destruction complete
aws_security_group_rule.consul_serv_to_client_udp: Destroying...
aws_security_group_rule.consul_agent_to_server_tcp: Destruction complete
aws_security_group_rule.consul_serv_to_client_tcp: Destruction complete
aws_security_group_rule.consul_serv_from_client_tcp: Destroying...
aws_security_group_rule.general_consul_agent_from_server_udp: Destroying...
aws_security_group_rule.general_consul_agent_to_server_udp: Destruction complete
aws_security_group.distribution: Destroying...
aws_security_group_rule.consul_serv_from_serv_tcp: Destruction complete
aws_security_group_rule.general_consul_agent_from_server_tcp: Destroying...
aws_security_group_rule.general_consul_agent_to_agent_udp: Destruction complete
aws_security_group_rule.consul_serv_to_serv_tcp: Destroying...
aws_security_group_rule.consul_serv_to_client_udp: Destruction complete
aws_security_group_rule.consul_serv_from_client_tcp: Destruction complete
aws_security_group_rule.general_consul_agent_to_agent_tcp: Destruction complete
aws_security_group_rule.general_consul_agent_from_agent_udp: Destruction complete
aws_security_group_rule.consul_serv_to_serv_tcp: Destruction complete
aws_security_group_rule.general_consul_agent_to_server_tcp: Destruction complete
aws_security_group_rule.general_consul_agent_from_server_udp: Destruction complete
aws_security_group_rule.general_consul_agent_from_server_tcp: Destruction complete
aws_security_group.general_access: Destroying...
aws_security_group.consul: Destroying...
Error applying plan:

4 error(s) occurred:

* aws_security_group.ansible: timeout while waiting for state to become '[success]'
* aws_security_group.distribution: timeout while waiting for state to become '[success]'
* aws_security_group.consul: timeout while waiting for state to become '[success]'
* aws_security_group.general_access: timeout while waiting for state to become '[success]'

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by vbichov** [comment]
Thank you very much for investing in this. This bug is a causing us a lot of problems...
Did you happen to try the "destroy -target=" on specific rules?

As I've mentioned above, later versions of Terraform have become less (?!) informative and apparently attempt to retry (in order to mitigate other timeout related issues maybe?) until they timeout and so, return a non-informative response.

(personally, I feel that this is a regression since the response is obviously false. This isn't a matter of timing-out, no amount of waiting will resolve this.)

**Comment by cloudynetwork** [comment]
Hi, I can also vouch for this issue. I am experiencing exactly the same thing. I'm using aws_security_group_rules throughout and the problem is simple. If security group A has a rule that points to security group B. Security group A cannot be deleted by terraform. The fix seems simple. Destroy the aws_security_group rules first..(or just the rules that point at other security groups). Then destroy the security groups.

**Comment by cloudynetwork** [comment]
Also, the error message has become less informative vs 0.6.9. At least in 0.6.9 you had the AWS message regarding the dependent security groups. In 0.6.11 I see only the message about state. 

* aws_security_group.linux_bastion_sg: timeout while waiting for state to become '[success]'
* aws_security_group.linux_administration_sg: timeout while waiting for state to become '[success]'

In order the destroy complete, all I have to do in my case is delete the single rule that prevents the SG from being deleted.

**Comment by phinze** [comment]
Hi @cloudynetwork - and @vbichov sorry for the trouble here! Terraform definitely orders destroys so that descendants are destroyed before ancestors, so it seems like there must be another issue at play.

It's looking like my tweak in #5460 might actually be fixing a race condition bug in addition to the error message shadowing. I'll retest the above repro w/ this branch and let you know if it should fix the behavior.

**Comment by cloudynetwork** [comment]
Hi @phinze I'm very much inclined to agree, I have only experienced this sporadically and especially in one or two situations a few weeks ago.

**Comment by phinze** [comment]
Retested the above config on master just now. With the following added to the top to make it work as a stand alone config:

```tf
variable "farm_name" {
  default = "foo"
}
resource "aws_vpc" "docker" {
  cidr_block = "10.0.0.0/16"
}
```

The SG destroy still does seem unable to sort itself out.

```

* aws_security_group.ansible: DependencyViolation: resource sg-1ff89178 has a dependent object
        status code: 400, request id:
* aws_security_group.distribution: DependencyViolation: resource sg-1df8917a has a dependent object
        status code: 400, request id:
* aws_security_group.general_access: DependencyViolation: resource sg-1ef89179 has a dependent object
        status code: 400, request id:
* aws_security_group.consul: DependencyViolation: resource sg-10f89177 has a dependent object
        status code: 400, request id:
```

Looking into this further. Stay tuned!

**Comment by phinze** [comment]
Going to pass this off to @catsby for further investigation. :+1:

**Comment by catsby** [comment]
Looking into this now

**Comment by catsby** [comment]
Hey all – 

The core of the original issue at top is the mix/match of protocol numbers and names, e.g. `"6"` and `"tcp"`. According to [this link](https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml) they are interchangeable, however Terraform was not treating them so. 

I've sent #5881 to patch this. For now though, use `"tcp"` (or other protocol _names_) as a workaround, until that patch is merged and a new release is shipped

**Comment by catsby** [comment]
I merged #5881 so I'm closing this; let me know if you have any questions!

**Comment by catsby** [comment]
#5881 was merged, so I'm going to close this for now. Thanks! Let me know if you have any other questions
